osu.Desktop/Updater/VelopackUpdateManager.cs

```diff
@@ -0,0 +1,127 @@
+        {
+            // should we schedule a retry on completion of this check?
+            bool scheduleRecheck = true;
+
+            try
+            {
+                // Avoid any kind of update checking while gameplay is running.
+                if (localUserInfo?.IsPlaying.Value == true)
+                    return false;
+
+                var info = await updateManager.CheckForUpdatesAsync().ConfigureAwait(false);
+
+                // Handle no updates available.
+                if (info == null)
+                {
+                    // If there's no updates pending restart, bail and retry later.
+                    if (!updateManager.IsUpdatePendingRestart) return false;
+
+                    // If there is an update pending restart, show the notification to restart again.
+                    notificationOverlay.Post(new UpdateApplicationCompleteNotification
+                    {
+                        Activated = () =>
+                        {
+                            restartToApplyUpdate();
+                            return true;
+                        }
+                    });
+                    return true;
+                }
+
+                scheduleRecheck = false;
+
+                if (notification == null)
+                {
+                    notification = new UpdateProgressNotification
+                    {
+                        CompletionClickAction = restartToApplyUpdate,
+                    };
+
+                    Schedule(() => notificationOverlay.Post(notification));
+                }
+
+                notification.StartDownload();
+
+                try
+                {
+                    await updateManager.DownloadUpdatesAsync(info, p => notification.Progress = p / 100f).ConfigureAwait(false);
+
+                    notification.State = ProgressNotificationState.Completed;
+                }
+                catch (Exception e)
+                {
+                    // In the case of an error, a separate notification will be displayed.
+                    notification.FailDownload();
+                    Logger.Error(e, @"update failed!");
+                }
+            }
+            catch (Exception e)
+            {
+                // we'll ignore this and retry later. can be triggered by no internet connection or thread abortion.
+                scheduleRecheck = true;
+                Logger.Error(e, @"update check failed!");
+            }
+            finally
+            {
+                if (scheduleRecheck)
+                {
+                    // check again in 30 minutes.
+                    Scheduler.AddDelayed(() => Task.Run(async () => await checkForUpdateAsync().ConfigureAwait(false)), 60000 * 30);
+                }
+            }
+
+            return true;
+        }
+
+        private bool restartToApplyUpdate()
+        {
+            updateManager.WaitExitThenApplyUpdates(null);
```

Comment:
Really need to provide an UpdateInfo here instead of null, the same one you passed to `DownloadUpdatesAsync`.

Reply:
@caesay how should we handle the case where `UpdateInfo` came back null but there's a pending update?

![JetBrains Rider 2024-09-02 at 08 14 41](https://github.com/user-attachments/assets/104248ce-fb09-4148-b99b-b3cc3aa8ca72)

Reply:
Based on the xmldoc it seems like passing null is fine in this case "The target release to apply. Can be left null to auto-apply the newest downloaded release."

Thanks for good documentation!

Reply:
Applied in https://github.com/ppy/osu/pull/28743/commits/cd9b82253e4639d3a94dc049244c9ccbcc59ea47.

Reply:
Maybe I should refactor `IsUpdatePendingRestart` to return a `VelopackAsset` so it can be passed in. Null will probably work most of the time but it's not best practice.

If you pass null, Velopack doesn't know for sure what update you actually want to apply, so it will just go searching in the packages directory for the latest version. Sometimes this is not what you want. For example, Velopack supports updating to older versions now - both explicitly and implicitly eg. it can automatically downgrade if you've retracted a release from your feed because it has breaking bugs as long as the `UpdateOptions.AllowVersionDowngrade` is set to true. 

So if you are downgrading a version but have not provided it to the apply/restart then you end up not installing it.

Also, DownloadUpdates should be a no-op if the release is already downloaded. 

My suggestion is you should not need to check IsUpdatePendingRestart except for on startup (but VelopackApp already does this for you by default.. see SetAutoApply there). 

You can just CheckForUpdates, if it's not null, call DownloadUpdates and then cache the UpdateInfo in a local field. If you have the local field non-null when a restart is requested then you know what to pass to Velopack.. If your local UpdateInfo is null, then you've not downloaded a new version yet.

Reply:
> see SetAutoApply

Hmm, this may be something we want to turn off to avoid the game hanging with no visible progress on startup for 10 seconds or more, I guess? This is going to block startup while the diff is applied correct?

Reply:
It does restart and apply on startup yes. On windows there is a progress dialog, nothing for mac/linux but I'm working on that. I recently just finished a cross-platform dialog library https://github.com/velopack/xdialog-rs but it has not yet been integrated to Velopack.

Reply:
On a faster PC this should only take a second or two, since it's just unzipping the nupkg. On a slower PC it's a bit unfortunate without the progress bar there.

Reply:
> You can just CheckForUpdates, if returns not null, call DownloadUpdates and then cache the UpdateInfo in a local field. If you have the local field non-null when a restart is requested then you know what to pass to Velopack.. If your local UpdateInfo is null, then you've not downloaded a new version yet.

Could there be a case where a second check-for-updates finds a newer update, where we'd want to use the newer `UpdateInfo`? Or is that not something that can happen in the update flow?

Reply:
It depends on how you implement things. CheckForUpdates will always return the latest update available. If you have already downloaded an update and another release has happened you can do another DownloadUpdates to grab the fresh version again.

Reply:
Okay good to know, thanks.

With my latest changes on this branch I bypassed that, but we'll probably add back support for it for fun.